node/src/utils/token.rs

```diff
@@ -0,0 +1,138 @@
+use std::fmt::{self, Debug};
+
+use tokio_util::sync::CancellationToken;
+
+#[derive(Clone)]
+pub(crate) struct Token {
+    token: CancellationToken,
+}
+
+impl Token {
+    pub(crate) fn new() -> Token {
+        Token {
+            token: CancellationToken::new(),
+        }
+    }
+
+    /// Trigger the event
+    pub(crate) fn trigger(&self) {
+        self.token.cancel();
+    }
+
+    /// Returns a guard that will trigger the event on drop.
+    pub(crate) fn trigger_drop_guard(&self) -> TokenTriggerDropGuard {
+        TokenTriggerDropGuard {
+            token: Some(self.token.clone()),
+        }
+    }
+
+    /// Returns if event is triggered or not.
+    #[allow(dead_code)]
+    pub(crate) fn is_triggered(&self) -> bool {
+        self.token.is_cancelled()
+    }
+
+    /// Returns when the event is triggered. If it was triggered before,
```

Comment:
```suggestion
    /// Returns when the token is triggered. If it was triggered before,
```
I think it's touch clearer when talking about just the token, without involving potential events user might stack on top of it.